---
title: CELLO-Seq Sarlacc pipeline - deduplication
author: Rebecca Berrens, Andrian Yang, Aaron Lun, Florian Bieberich
output:
  BiocStyle::html_document:
    toc_float: true
    titlecaps: false
---

```{r, echo=FALSE, results="hide", message=FALSE}
require(knitr)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

```{r setup, echo=FALSE, message=FALSE}
library(sarlacc)
```

# Introduction

This series of Rmarkdown files forms the CELLO-Seq sarlacc data processing pipeline, which utilises the `sarlacc` R package as its base.

After UMI grouping, we can either perform errorcorrection or deduplication to collapse each read groups into a single read.
In the deduplicate step, we randomly select a read from each group to select as the representative read.

# Reading files

We read in the read groups along with the RDS file containing the sequence information for all reads.

```{r}
my.sample_fastq <- readRDS(paste0(gsub(".transcript", "", INDEX), ".rds"))
groups <- readRDS(paste0(INDEX, ".groups.", CHUNK, ".rds", sep=""))
```

# Deduplication of reads

We then pick a random read from each read group as the representative read.

```{r}
resample <- function(x, ...) x[sample.int(length(x), ...)]
selected <- lapply(groups, resample, 1)
```

# Writing out reads

We read in the sequence of each deduplicated read using the `realizeReads` function.

```{r}
my.sample_fastq <- compact(my.sample_fastq[unlist(selected),])
read.seq <- compact(realizeReads(my.sample_fastq))
```

We rename the reads for consistency with the consensus step, following the format of `sampleName_commonName_chunkID_readID`^[chunkID corresponds to the ID of the read groups chunks being processed (see grouping.Rmd for further explanation), while readID corresponds to the unique ID of the read/read group within the corresponding chunks.].

```{r}
varnames <- c("ConsRead")
names(cons.out) <- paste0(INDEX, "_", varnames, "_", CHUNK, "_", seq_along(cons.out))

read.seq
```

We now output the deduplicated reads to a FASTQ file.

```{r}
outfile <- paste0(INDEX, "_", CHUNK, "_deduplicated.fastq")
writeQualityScaledXStringSet(read.seq, outfile)
```

# Session information

```{r}
sessionInfo()
```